helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -202,6 +212,26 @@ protected void handleEvent(NotificationContext event) {
     }
   }
 
+  class RefreshTask implements Runnable {
+    @Override
+    public void run() {
+      try {
+        long currentTime = System.currentTimeMillis();
+        if (_lastEventTime + _periodicRefreshInterval <= currentTime) {
+          NotificationContext changeContext = new NotificationContext(_manager);
+          changeContext.setType(NotificationContext.Type.PERIODIC_REFRESH);
+          changeContext.setChangeType(_changeType);
+          enqueueTask(changeContext);
+        } else {
+          long remainingTime = _lastEventTime + _periodicRefreshInterval - currentTime;
+          Thread.sleep(remainingTime);
```

Comment:
You need to do the refresh check again here, otherwise a whole new period/cycle will pass without refresh. :)

Reply:
Didn't think of that! Thanks for catching this. I think I would prefer my previous method then, the framework handles it better. And I searched, to cancel a future and to schedule another one, is merely to remove a task from workQueue and create another task and put it into workQueue, so it won't be too costly, especially if we are doing it for relatively long interval (20, 30 minutes).

Reply:
Ok. I personally liked this approach because it's cleaner, but I'll let you be the judge of that.